File: regex_extracting/asp_version/Extracting/Color.cs
/*
 * author : xm
 * date : 2013/9/3
 * */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Web;

namespace Ch2R.Code.Extracting
{
    public class Color : ParameterBase
    {
        public Color(Areas.Chat.Models.UserSession userSession, string sentence)
            : base(userSession, sentence)
        {
            name = "颜色";

            #region 初始化正则表达式
            // 描述正则表达式
            StatementRegexs = new string[] {
                @"颜色"
            };
            // 值正则表达式
            ValueRegexs = new string[] {
                @"樱花粉|中国红|湛清蓝|现代银|柠檬绿|诱惑紫|珍珠白|深邃黑|粉红|红|蓝|黑|白|紫|土豪金|金|摩卡啡",
                @"随意|随便|都可以|其他|别的"
            };
            #endregion

            RegexProcess();

        }

        public override void Format()
        {
            for (int i = 0; i < list.Count; ++i)
            {
                if (Regex.IsMatch(list[i].rawData, @"随意|随便|都可以"))
                {
                    list[i].normalData = "";
                }
                else if (Regex.IsMatch(list[i].rawData, @"别的|其他的"))
                {
                    if (userSession.liveTable.Parameter["颜色"] != null)
                    {
                        list[i].normalData = (String)userSession.liveTable.Parameter["颜色"];
                        list[i].isNegative = true;
                    }
                    else
                    {
                        /**
                         * 仍有其他价格为null的情况
                         * 别的未测试
                         * */
                    }
                }
                else
                {
                    list[i].normalData = list[i].rawData;
                }
                /* 这里处理一下否定情况
                 * 做法先只是简单的判断原句中信息元数据之前有没有出现否定语
                 * 比如用户说：我不要红色的
                 * */
                if (Regex.IsMatch(sentence.Substring(0, list[i].leftIndex), @"(不)|(非(?!常))|其他|其它|别的"))
                {
                    list[i].isNegative = true;
                }
            }
        }

        public override InfoBlock ToInfoBlock()
        {
            List<InfoUnit> L = new List<InfoUnit>();
            foreach (InfoMetadata im in list)
            {
                InfoUnit iu = new InfoUnit();
                iu.normalData = im.normalData;
                iu.rawData = im.rawData;
                iu.isNegative = im.isNegative;
                L.Add(iu);
            }
            InfoBlock ib = new InfoBlock(name, "or", L);

            return ib;
        }
    }
}

File: regex_extracting/extracting/regex_lib/Color.py
# encoding=utf8
"""
    Author:  'jdwang'
    Date:    'create date: 2017-01-13'; 'last updated date: 2017-01-13'
    Email:   '383287471@qq.com'
    Describe:
"""
from __future__ import print_function
from regex_extracting.extracting.common.regex_base import RegexBase

__version__ = '1.2'


class Color(RegexBase):
    name = '颜色'

    def __init__(self,sentence):

        # 要处理的输入句子

        self.sentence = sentence

        # region 1 初始化正则表达式
        # 描述正则表达式
        self.statement_regexs = [
            '颜色'
        ]
        # 值正则表达式
        self.value_regexs = [
            '樱花粉|中国红|湛清蓝|现代银|柠檬绿|诱惑紫|珍珠白|深邃黑|粉红|红|蓝|黑|白|紫|土豪金|金|摩卡啡',
                '随意|随便|都可以|其他|别的'
        ]
        # endregion
        super(Color, self).__int__()
        self.regex_process()


if __name__ == '__main__':
    price = Color('价不高')
    for info_meta_data in price.info_meta_data_list:
        print('-'*80)
        print(str(info_meta_data))